PYME/recipes/kinetics_recipes.py

```diff
@@ -0,0 +1,428 @@
+
+        oncol = self.onTimesColName
+        offcol = self.offTimesColName
+
+        inp = namespace[self.inputName]
+
+        if 'mdh' in dir(inp):
+            mdh = inp.mdh
+            cyctime = mdh.getEntry('Camera.CycleTime')
+        else:
+            cyctime = self.exposure_time
+
+        ds = tabular.mappingFilter(inp)
+
+        off_fit_limit_index = np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int')
+
+        offtimeslimit = np.round(self.minOffTimeSecondsToFit/cyctime).astype('int')
+
+        onbins = np.arange(1, max(ds[oncol])+1, 1)
+        offbins = np.arange(offtimeslimit, off_fit_limit_index, np.round(self.offTimesBinSizeInSeconds/cyctime).astype('int'))
+
+
+        n_on, bins_on = np.histogram(ds[oncol], bins=onbins)
+        n_off, bins_off = np.histogram(ds[offcol], bins=offbins)
+
+        bins_on = bins_on[:-1]
+        bins_off = bins_off[:-1]
+
+        res_off = FitModelWeighted(eimod, [n_off[0]*self.off_init_condition_multiple, 1], n_off, 1. / (np.sqrt(n_off) + 1), bins_off * cyctime)
+
+        res_on = FitModelWeighted(eimod, [n_on[1], 1], n_on[1:], 1. / (np.sqrt(n_on[1:]) + 1), bins_on[1:] * cyctime)
+
+        pylab.figure()
+
+        pylab.bar(bins_on*cyctime, n_on, width=np.diff(bins_on)[0]*cyctime, alpha=0.4, fc=colours[0])
+        pylab.plot(np.linspace(bins_on[0], max(bins_on), max(bins_on) * 10) * cyctime,
+                   eimod(res_on[0], np.linspace(bins_on[0], max(bins_on), max(bins_on) * 10) * cyctime), colours[0],
+                   lw=3)
+
+        pylab.ylabel('Events')
+        pylab.xlabel('Event Duration [s]')
+        pylab.xlim((0, np.round(n_on[-1]*1.2).astype('int')))
+        pylab.title('Blink Fluorescence Durations')
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_on[0][1]), size=18, color=colours[1])
+        pylab.show()
+
+
+        pylab.figure()
+
+        pylab.xlim(right=np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int'))
+        pylab.bar(bins_off * cyctime, n_off, width=np.diff(bins_off)[0] * cyctime, alpha=0.3, fc=colours[2])
+        pylab.plot(np.linspace(bins_off[0], max(bins_off), max(bins_off) * 10) * cyctime,
+                   eimod(res_off[0], np.linspace(1, max(bins_off), max(bins_off) * 10) * cyctime),
+                   colours[3], lw=3, alpha=.8)
+        pylab.ylabel('Events')
+        pylab.xlabel('Event Duration [s]')
+        pylab.ylim((.9, pylab.ylim()[1]))
+        pylab.title('Blink dark times')
+
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_off[0][1]), size=18, color=colours[1])
+
+        pylab.show()
+
+        powers = np.array([res_on[0][1], res_off[0][1]])
+        times = np.array([cyctime, (max(ds[oncol]).astype(float)+1)*cyctime, (max(ds[offcol]).astype(float)+1)*cyctime])
+
+        output_list = [powers, times, n_on, n_off]
+
+        namespace[self.outputEmphistData] = output_list
+
+@register_module('save_EmpHistJSON')
+class save_EmpHistJson(ModuleBase):
+    """
```

Comment:
I'm assuming you went json here since most tabular types have 1 x N columns only. We do have ragged datastructures, see `PYME.IO.ragged`. Those have a to_json method which might do what you want.

If you want to run this without a GUI, e.g. batch processed, on the cluster, etc., you'd want to use an output module in `recipes/output.py`. Unfortunately we don't have one for json at the moment, which we should maybe consider. That said, you do just have 2 length N columns and a couple constants, and those columns are nothing you'd care to look at in a human-readable uncompressed file, so I would maybe suggest instead going the hdf route and using the metadata to store scalars. There should be an example in the spherical harmonic shell stuff. There are maybe a couple ways to do this, but would be nice to use one of the existing data structures, either in ragged or tabular.